linera-execution/src/system.rs

```diff
@@ -1117,6 +1142,169 @@ where
+        let chain_state = bincode::serialize(&chain_state)
+            .map_err(|_| SystemExecutionError::SerializationFailed("chain_state".into()))?;
+        let zstore_view = zstore.to_view();
+        let zstore_view = bincode::serialize(&zstore_view)
+            .map_err(|_| SystemExecutionError::SerializationFailed("zstore_view".into()))?;
+
+        return Ok((chain_proofs, chain_state, zstore_view));
+    }
+
+    pub async fn microchain_transition(
+        &mut self,
+        chain_proof_id: BlobId,
+        chain_proofs: Vec<u8>,
+        chain_state: Vec<u8>,
+        zstore_view: Vec<u8>,
+    ) -> Result<(Vec<u8>, Vec<u8>, Vec<u8>), SystemExecutionError> {
+        let chain_proof_bytes = self.read_blob_content(chain_proof_id).await?.into_bytes();
+        let chain_proof: ChainProof = bincode::deserialize_from(&chain_proof_bytes[..])
+            .map_err(|_| SystemExecutionError::DeserializationFailed("chain_proof".into()))?;
+        let mut chain_proofs: Vec<OpaqueChainProof> = bincode::deserialize_from(&chain_proofs[..])
+            .map_err(|_| SystemExecutionError::DeserializationFailed("chain_proofs".into()))?;
+        let chain_state: ChainState = bincode::deserialize_from(&chain_state[..])
+            .map_err(|_| SystemExecutionError::DeserializationFailed("chain_state".into()))?;
+        let zstore_view: ZStoreView<BabyBear> = bincode::deserialize_from(&zstore_view[..])
+            .map_err(|_| SystemExecutionError::DeserializationFailed("zstore_view".into()))?;
+        let mut zstore = ZStore::from_view(zstore_view, lurk_hasher());
+
+        let ChainProof {
+            crypto_proof,
+            call_args,
+            next_chain_result,
+            next_callable,
+        } = chain_proof;
+
+        let next_chain_result_zptr = {
+            if next_chain_result.is_flawed(&mut zstore) {
+                return Err(SystemExecutionError::FlawedData(format!(
+                    "next_chain_result ({:?})",
+                    next_chain_result.zptr.digest
+                )));
+            }
+            next_chain_result.zptr
+        };
+
+        let next_callable_zptr = match &next_callable {
+            CallableData::Comm(comm_data) => {
+                if comm_data.payload_is_flawed(&mut zstore) {
+                    return Err(SystemExecutionError::FlawedData(format!(
+                        "comm_data ({:?})",
+                        comm_data.payload.digest
+                    )));
+                }
+                comm_data.commit(&mut zstore)
+            }
+            CallableData::Fun(lurk_data) => {
+                if lurk_data.is_flawed(&mut zstore) {
+                    return Err(SystemExecutionError::FlawedData(format!(
+                        "lurk_data ({:?})",
+                        lurk_data.zptr.digest
+                    )));
+                }
+                lurk_data.zptr
+            }
+        };
```

Comment:
Just as a small mention/heads up, this handling of `next_callable_zptr` above will likely need to be extended. The callable can also be a bignum or a symbol, and we're likely to want to forbid other types of callables. No need to change anything in particular in this PR